File: protocol/src/test/csharp/CsProtocol/Packet/SimpleObject.cs
using System;
using System.Collections.Generic;
using CsProtocol.Buffer;

namespace CsProtocol
{
    
    public class SimpleObject : IProtocol
    {
        public int c;
        public bool g;

        public static SimpleObject ValueOf(int c, bool g)
        {
            var packet = new SimpleObject();
            packet.c = c;
            packet.g = g;
            return packet;
        }


        public short ProtocolId()
        {
            return 104;
        }
    }


    public class SimpleObjectRegistration : IProtocolRegistration
    {
        public short ProtocolId()
        {
            return 104;
        }

        public void Write(ByteBuffer buffer, IProtocol packet)
        {
            if (buffer.WritePacketFlag(packet))
            {
                return;
            }
            SimpleObject message = (SimpleObject) packet;
            buffer.WriteInt(message.c);
            buffer.WriteBool(message.g);
        }

        public IProtocol Read(ByteBuffer buffer)
        {
            if (!buffer.ReadBool())
            {
                return null;
            }
            SimpleObject packet = new SimpleObject();
            int result0 = buffer.ReadInt();
            packet.c = result0;
            bool result1 = buffer.ReadBool();
            packet.g = result1;
            return packet;
        }
    }
}

File: protocol/src/test/ecmascript/zfooes/packet/SimpleObject.mjs

class SimpleObject {
    c = 0; // number
    g = false; // boolean
}

export class SimpleObjectRegistration {
    protocolId() {
        return 104;
    }

    write(buffer, packet) {
        if (packet === null) {
            buffer.writeInt(0);
            return;
        }
        buffer.writeInt(-1);
        buffer.writeInt(packet.c);
        buffer.writeBoolean(packet.g);
    }

    read(buffer) {
        const length = buffer.readInt();
        if (length === 0) {
            return null;
        }
        const beforeReadIndex = buffer.getReadOffset();
        const packet = new SimpleObject();
        const result0 = buffer.readInt();
        packet.c = result0;
        const result1 = buffer.readBoolean(); 
        packet.g = result1;
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length);
        }
        return packet;
    }
}

export default SimpleObject;